### PR TITLE
Python wrapper for omniscidb engine

### PR DESCRIPTION
**Pull request by igalink**

### CODE REVIEW

File: Embedded/DBEngine.cpp [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+    } catch (std::exception const& e) {
+      std::cerr << "DBE:init: " << e.what() << std::endl;
+    } catch (...) {
+      std::cerr << "DBE:init: Unknown exception" << std::endl;
+    }
+    return false;
+  }
+
+  void reset() {
+    std::cout << "DBE:reset" << std::endl;
```
Comment (anton-malakhov) [inline Embedded/DBEngine.cpp:159]: is this a debugging artifact?

File: Embedded/DBEngine.cpp [hunk 2]
```diff
@@ -159,27 +293,29 @@ class DBEngineImpl : public DBEngine {
...
-DBEngine* DBEngine::create(std::string path) {
-  return new DBEngineImpl(path);
+DBEngine* DBEngine::create(std::string path, int calcite_port, bool enable_columnar_output) {
+  g_enable_columnar_output = enable_columnar_output;
+  return new DBEngineImpl(path, calcite_port);
 }
 
 /** DBEngine downcasting methods */
+
 inline DBEngineImpl* getImpl(DBEngine* ptr) {
```
Comment (fexolm) [inline Embedded/DBEngine.cpp:308]: I see, that this is out of the scope of this pr and it's already merged in master, but I'm not sure that downcasting is really necessary here. This inheritance looks too overcomplicated.

Comment (fexolm) [reply Embedded/DBEngine.cpp:308]: Also, please avoid C style cast, as they mask what kind of cast they really do

Comment (fexolm) [reply Embedded/DBEngine.cpp:308]: Maybe what you wanted to implement here is just PIMPL. If so, you don't need inheritance at all.
See https://en.cppreference.com/w/cpp/language/pimpl

**Comment by fexolm** [comment]
Please rebase this commit to develop branch, as you have some commits from modin here.
We don't really need that changes to be merged to master, but if you want this pr to be merged, please rebase it.

**Comment by fexolm** [comment]
Also, I agree with Anton, it is really unexpected behaviour from the library to just print the errors to stderr. You may better not handle exceptions in your code at all, to let the user handle them. Please rework the way how you handle exceptions.

File: Embedded/DBETypes.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    auto value = boost::get<NullableString>(scalar_value);
+    bool is_null = !value || boost::get<void*>(value);
+    if (is_null) {
+      return "Empty";
+    } else {
+      auto value_notnull = boost::get<std::string>(value);
+      return *value_notnull;
+    }
+  }
+  return "Out of range";
```
Comment (fexolm) [inline Embedded/DBETypes.cpp:85]: Does it expected to this method to return that hardcoded strings?

File: Embedded/DBEngine.cpp [hunk 4]
```diff
@@ -16,138 +16,272 @@
...
+        std::cout << "Columnar format enabled" << std::endl;
+      }
+    } else {
+      std::cerr << "DBEngine initialization failed" << std::endl;
+    }
+  }
+
+  bool init(std::string base_path, int calcite_port) {
+    std::cout << "DBE:init(" << base_path << ", " << calcite_port << ")" << std::endl;
+    if (query_runner_) {
```
Comment (fexolm) [inline Embedded/DBEngine.cpp:117]: why can not we reuse code from QueryRunner::init? Maybe not exactly here, but somewhere? It looks like QueryRunner does exactly what DBEngine does. We have also have experimented and were able to run DDL and DML with just a few lines of code in DB engine. I'm wondering why all that borrowed code is necessary.

File: Embedded/CMakeLists.txt [hunk 5]
```diff
@@ -1,3 +1,38 @@
...
+target_link_libraries(DBEngine QueryRunner QueryEngine DataMgr ${Arrow_LIBRARIES} ${Boost_LIBRARIES} ${CMAKE_DL_LIBS} ${TBB_LIBS})
+
+set(INSTALL_PATH "${CMAKE_INSTALL_PREFIX}")
+install(TARGETS DBEngine DESTINATION ${INSTALL_PATH})
+
+add_executable(EmbeddedDbTest EmbeddedDbTest.cpp)
+target_link_libraries(EmbeddedDbTest DBEngine)
+
+find_program(PYTHON "python3")
+IF(PYTHON)
```
Comment (fexolm) [inline Embedded/CMakeLists.txt:12]: you don't need if to be caps in cmake

File: DataMgr/ForeignStorage/ArrowCsvForeignStorage.cpp [hunk 6]
```diff
@@ -528,18 +528,25 @@ void ArrowCsvForeignStorage::registerTable(Catalog_Namespace::Catalog* catalog,
                                            const TableDescriptor& td,
                                            const std::list<ColumnDescriptor>& cols,
                                            Data_Namespace::AbstractBufferMgr* mgr) {
+  const DataframeTableDescriptor* df_td =
+      dynamic_cast<const DataframeTableDescriptor*>(&td);
+  bool isDataframe = df_td ? true : false;
+  if (!isDataframe) {
+    df_td = new DataframeTableDescriptor(td);
```
Comment (fexolm) [inline DataMgr/ForeignStorage/ArrowCsvForeignStorage.cpp:535]: can you use smart pointers here and in all other places where you use `new`? Have you checked your code with the address sanitizer?

**Comment by fexolm** [comment]
Also, as far as I see, there is a lot of allocations, which are never deallocated, like 
```cpp
        cursors_.emplace_back(new CursorImpl(rs, col_names, data_mgr_));
```
I haven't find the place where this memory was released.
Have you checked you code with the address sanitizer?
Please avoid using `new` allocations, use RAII alternative instead in all places where possible.

**Pull request closed without merging**